scipy/linalg/tests/test_decomp.py

```diff
@@ -351,6 +351,30 @@ def test_shape_mismatch(self):
         assert_raises(ValueError, eig, A, B)
         assert_raises(ValueError, eig, B, A)
 
+    def test_gh_11577(self):
+        # https://github.com/scipy/scipy/issues/11577
+        # `A - lambda B` should have 4 and 8 among the eigenvalues, and this
+        # was apparently broken on some platforms
+        A = np.array([[12.0, 28.0, 76.0, 220.0],
+                      [16.0, 32.0, 80.0, 224.0],
+                      [24.0, 40.0, 88.0, 232.0],
+                      [40.0, 56.0, 104.0, 248.0]], dtype='float64')
+        B = np.array([[2.0, 4.0, 10.0, 28.0],
+                      [3.0, 5.0, 11.0, 29.0],
+                      [5.0, 7.0, 13.0, 31.0],
+                      [9.0, 11.0, 17.0, 35.0]], dtype='float64')
+
+        D, V = eig(A, B)
+
+        # The problem is ill-conditioned, and two other eigenvalues
+        # depend on ATLAS/OpenBLAS version, compiler version etc
+        # see gh-11577 for discussion
+        with np.testing.suppress_warnings() as sup:
+            # isclose chokes on inf/nan values
+            sup.filter(RuntimeWarning, "invalid value encountered in multiply")
+            assert np.isclose(D, 4.0, atol=1e-14).any()
+            assert np.isclose(D, 8.0, atol=1e-14).any()
```

Comment:
Any reason not to use the built-in testing assertion? This works for me locally at least:

```diff
--- a/scipy/linalg/tests/test_decomp.py
+++ b/scipy/linalg/tests/test_decomp.py
@@ -369,11 +369,7 @@ class TestEig:
         # The problem is ill-conditioned, and two other eigenvalues
         # depend on ATLAS/OpenBLAS version, compiler version etc
         # see gh-11577 for discussion
-        with np.testing.suppress_warnings() as sup:
-            # isclose chokes on inf/nan values
-            sup.filter(RuntimeWarning, "invalid value encountered in multiply")
-            assert np.isclose(D, 4.0, atol=1e-14).any()
-            assert np.isclose(D, 8.0, atol=1e-14).any()
+        assert_allclose(D[:2], [4.0, 8.0], atol=1e-14)
```

I think that may give slicker error messages if it ever fails alongside graceful non-finite handling.

Apart from that, looks solid.

Reply:
My original reasoning was 1) I don't remember if assert_allclose is gracefully handling nans and infs, and 2) I don't know if the order is guaranteed, whether two "other" values are always in the 3rd and 4th places.

That said, let's declare both concerns to be FUD until proven real. So the PR is updated with your patch, in a separate commit to simplify reverting if it ever comes to that :-).

Reply:
Okay, here it is: locally the order is `[8.0, 4.0, rest]` and CI failures are because it is `[4.0, 8.0, rest]`. Typically I'd expect that I'd need to sort the eigenvalues before comparing them because the ordering from LAPACK is not guaranteed. Here sorting is not an option because of "rest". OK to revert to the previous manual version?

Reply:
> locally the order is [8.0, 4.0, rest]

Ok, so even your local order doesn't match my local order from above.

That's annoying. I see `eigvals` explicitly documents `w` as:

> The eigenvalues, each repeated according to its multiplicity but not in any specific order. 

While the function you use here (`eig`) doesn't actually say that in the docs. Reverting makes sense then I guess. There are other blemishes in the API too, like the left eigenvector not being normalized while the right one is normalized (a fact that wasn't even documented until recently).

Reply:
> There are other blemishes in the API too

ISTM the root is that `eig` tries to be too helpful in too many ways. Cramming two rather different things into one function (regular and generalized eigenvalue problem, for starters). Then it tries to go beyond what LAPACK actually offers (sorting? normalization?). Water under the bridge now anyway.